pennylane_lightning/lightning_gpu/_state_vector.py

```diff
@@ -0,0 +1,337 @@
+
+            # TODO
+            # Create an implementation in the C++ backend and binding to be able
+            # to allocate memory for a new statevector and copy the data
+            # from an external state vector.
+            # state_data = allocate_aligned_array(state.size, np.dtype(self.dtype), True)
+            # state.getState(state_data)
+            # state = state_data
+
+        state = self._asarray(state, dtype=self.dtype)  # this operation on host
+        output_shape = [2] * self._num_local_wires
+
+        if len(device_wires) == self.num_wires and Wires(sorted(device_wires)) == device_wires:
+            # Initialize the entire device state with the input state
+            if self.num_wires == self._num_local_wires:
+                self.syncH2D(np.reshape(state, output_shape))
+                return
+            local_state = np.zeros(1 << self._num_local_wires, dtype=self._dtype)
+            self._mpi_handler.mpi_manager.Scatter(state, local_state, 0)
+            self.syncH2D(np.reshape(local_state, output_shape))
+            return
+
+        # generate basis states on subset of qubits via the cartesian product
+        basis_states = np.array(list(product([0, 1], repeat=len(device_wires))))
+
+        # get basis states to alter on full set of qubits
+        unravelled_indices = np.zeros((1 << len(device_wires), self.num_wires), dtype=int)
+        unravelled_indices[:, device_wires] = basis_states
+
+        # get indices for which the state is changed to input state vector elements
+        ravelled_indices = np.ravel_multi_index(unravelled_indices.T, [2] * self.num_wires)
+
+        # set the state vector on GPU with the unravelled_indices and their corresponding values
+        self._qubit_state.setStateVector(
+            ravelled_indices, state, use_async
+        )  # this operation on device
+
+    def _apply_lightning_controlled(self, operation):
+        """Apply an arbitrary controlled operation to the state tensor.
+
+        Args:
+            operation (~pennylane.operation.Operation): controlled operation to apply
+
+        Returns:
+            None
+        """
+        state = self.state_vector
+
+        control_wires = list(operation.control_wires)
+        control_values = operation.control_values
+        name = operation.name
+        # Apply GlobalPhase
+        inv = False
+        param = operation.parameters[0]
+        wires = self.wires.indices(operation.wires)
+        matrix = global_phase_diagonal(param, self.wires, control_wires, control_values)
+        state.apply(name, wires, inv, [[param]], matrix)
+
+    def _apply_lightning_midmeasure(self):
+        """Execute a MidMeasureMP operation and return the sample in mid_measurements.
+
+        Args:
+
+        Returns:
+            None
+        """
+        raise DeviceError("LightningGPU does not support Mid-circuit measurements.")
```

Comment:
Codecov complains

Reply:
MCM test is quite complicated to extend a test for non-supported devices. Perhaps, we can explore this in the future.